bundle/src/main/java/com/adobe/acs/commons/packagegarbagecollector/PackageGarbageCollectionJob.java

```diff
@@ -0,0 +1,126 @@
+import java.io.IOException;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.Period;
+import java.time.format.DateTimeFormatter;
+import java.time.format.FormatStyle;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.List;
+
+import static com.adobe.acs.commons.packagegarbagecollector.PackageGarbageCollectionScheduler.*;
+
+@Component(
+    service = JobConsumer.class,
+    immediate = true,
+    property = { JobConsumer.PROPERTY_TOPICS + "=" + PackageGarbageCollectionScheduler.JOB_TOPIC })
+public class PackageGarbageCollectionJob implements JobConsumer {
+    private static final Logger LOG = LoggerFactory.getLogger(PackageGarbageCollectionJob.class);
+
+    private static final String SERVICE_USER = "package-garbage-collection";
+
+    @Reference
+    Packaging packaging;
+
+    @Reference
+    ResourceResolverFactory resourceResolverFactory;
+
+    @Override
+    public JobResult process(Job job) {
+        Session session = null;
+        String groupName = job.getProperty(GROUP_NAME, String.class);
+        Integer maxAgeInDays = job.getProperty(MAX_AGE_IN_DAYS, Integer.class);
+        LOG.debug("Job Configuration: ["
+                + "Group Name: {}, "
+                + "Service User: {}, "
+                + "Age of Package {} days,]", groupName, SERVICE_USER, maxAgeInDays);
+
+        try (ResourceResolver resourceResolver = resourceResolverFactory.getServiceResourceResolver(
+            Collections.singletonMap(ResourceResolverFactory.SUBSERVICE, SERVICE_USER))) {
+            session = resourceResolver.adaptTo(Session.class);
+            JcrPackageManager packageManager = packaging.getPackageManager(session);
+            List<JcrPackage> packages = packageManager.listPackages(groupName, false);
+
+            for (JcrPackage jcrPackage : packages) {
+                String packageDescription = getPackageDescription(jcrPackage);
+                if (isPackageOldEnough(jcrPackage, maxAgeInDays)) {
+                    packageManager.remove(jcrPackage);
+                    LOG.info("Deleted package {}", packageDescription);
+                } else {
+                    LOG.info("Not removing package because it's not old enough {}", packageDescription);
+                }
+            }
+        } catch (LoginException | RepositoryException | IOException e) {
+            LOG.error("Unable to clear packages due to {}", e.toString());
+            return JobResult.FAILED;
+        } finally {
+            if (session != null) {
+                session.logout();
+            }
+        }
+        LOG.info("Package Garbage Collector job finished");
+        return JobResult.OK;
+    }
+
+    private boolean isPackageOldEnough(JcrPackage jcrPackage, Integer maxAgeInDays) throws RepositoryException, IOException {
+        DateTimeFormatter formatter = DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM);
+        Period maxAge = Period.ofDays(maxAgeInDays);
```

Comment:
IMHO only package versions which are not currently installed should ever be eligible for GC

Reply:
Fair to a degree - I can see your point for the "latest version" of installed packages that are older than the defined configuration. I.e. - there a deployment cycle that for some reason happens to be longer than the the configured max age. I can see this happening perhaps during change freezes where a normal two week deployment. turns into 2 months. To a degree, this could be handled by careful configuration of the max age parameter....but I do agree, it might be good to never remove the latest one...

Or is this not what you were referring to?

Reply:
The latest one is not necessarily the one which is installed. Please put a guard in place that never removes the currently installed one (no matter whether it is the latest or some other version).  Here is some code example which should give you an idea how to implement such a check

```
/**
     * Verifies if a package is installed. A package is considered installed if it has a lastUnpacked property
     * and no other version of that package was installed at a more recent date.
     *
     * @param package
     * @return {@code true} if the package is installed
     */
    public boolean isInstalled(Collection<ContentPackage> installedPackages, Logger logger) {

        Optional<ContentPackage> lastInstalledPackageOptional = installedPackages.stream()
                .filter(this::isSameGroupAndName)
                .sorted(Collections.reverseOrder(Comparator.comparing(ContentPackage::getLastUnpacked)))
                .findFirst();
        if (!lastInstalledPackageOptional.isPresent()) {
            logger.info("Package " + getName() + " not found in any version");
            return false;
        }
        ContentPackage lastInstalledPackage = lastInstalledPackageOptional.get();
        
        if (lastInstalledPackage.getLastUnpacked() == null) {
            logger.info("Package " + getName() + " is present but has not been installed yet");
            return false;
        }
        
        boolean isInstalled = hasSamePid(lastInstalledPackage);
        if (!isInstalled) {
            logger.info("Package " + getName() + " was last installed with another version "
                    + lastInstalledPackage.getVersion() + " at " + DateTimeFormatter.ISO_DATE_TIME.format(lastInstalledPackage.getLastUnpacked()));
        }
        return isInstalled;
    }
```

Reply:
Added logic to cater for this guard.